cedar-policy-validator/src/type_error.rs

```diff
@@ -332,3 +338,71 @@ pub struct WrongCallStyle {
 pub struct FunctionArgumentValidationError {
     msg: String,
 }
+
+/// Contains more detailed information about an attribute access when it occurs
+/// on an entity type expression or on the `context` variable. Track a `Vec` of
+/// attributes rather than a single attribute so that on `principal.foo.bar` can
+/// report that the record attribute `foo` of an entity type (e.g., `User`)
+/// needs attributes `bar` instead of giving up when the immediate target of the
+/// attribute access is not a entity.
+#[derive(Debug, Hash, Eq, PartialEq)]
+pub(crate) enum AttributeAccess {
+    /// The attribute access is some sequence of attributes accesses eventually
+    /// targeting an EntityLUB.
+    EntityLUB(EntityLUB, Vec<SmolStr>),
+    /// The attribute access is some sequence of attributes accesses eventually
+    /// targeting the context variable.
+    Context(EntityUID, Vec<SmolStr>),
+    /// Other cases where we do not attempt to give more information about the
+    /// access. This includes any access on the `AnyEntity` type and on record
+    /// types other than the `context` variable.
+    Other(Vec<SmolStr>),
+}
+
+impl AttributeAccess {
+    pub(crate) fn from_expr(
+        req_env: &RequestEnv,
+        mut expr: &Expr<Option<Type>>,
```

Comment:
why `mut` expr?

Reply:
This just felt nicer to write as a loop rather than recursively, so I need to do the mutable update `expr = sub_expr`.